### PR TITLE
Timestamp replace compatibility

### LINKED ISSUE #15240: Timestamp.replace raises ValueError instead of TypeError when given wrong argument
#### Code Sample, a copy-pastable example if possible

```python

from pandas import Timestamp
Timestamp("2012").replace(apple=3)
```
#### Problem description

The reimplementation of Timestamp.replace (https://github.com/pandas-dev/pandas/commit/f8bd08e9c2fc6365980f41b846bbae4b40f08b83#diff-05de2950b2c86c9828531957e02d1b87L662) raises a ValueError when a keyword argument does not exist.
Previous implementation used datetime.replace that raises a TypeError when an argument is not valid.
This is a regression and it is probably better to keep API compatibility with the datetime.replace function (modulo extra functionalities).
For autodocumentation/autocomplete purposes, it could also be useful to have explicitly in the function signature the keywords, so instead of::

`  def replace(self, **kwds):
`
have::
 
`  def replace(self, year=None, month=None, day=None, hour=None, minute=None, second=None, microsecond=None, nanosecond=None, tzinfo=None)
`

#### Actual Output
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "pandas\tslib.pyx", line 715, in pandas.tslib.Timestamp.replace (pandas\tslib.c:14832)
ValueError: invalid name apple passed

#### Expected Output
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "pandas\tslib.pyx", line 715, in pandas.tslib.Timestamp.replace (pandas\tslib.c:14832)
TypeError: 'apple' is an invalid keyword argument for this function

#### Output of ``pd.show_versions()``

<details>
INSTALLED VERSIONS
------------------
commit: None
python: 3.5.2.final.0
python-bits: 64
OS: Windows
OS-release: 7
machine: AMD64
processor: Intel64 Family 6 Model 61 Stepping 4, GenuineIntel
byteorder: little
LC_ALL: None
LANG: None
LOCALE: None.None

pandas: 0.19.2
nose: None
pip: 9.0.1
setuptools: 27.2.0
Cython: None
numpy: 1.11.3
scipy: None
statsmodels: None
xarray: None
IPython: 5.1.0
sphinx: 1.5.2
patsy: None
dateutil: 2.6.0
pytz: 2016.10
blosc: None
bottleneck: None
tables: None
numexpr: None
matplotlib: None
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: None
httplib2: None
apiclient: None
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: 2.9.4
boto: None
pandas_datareader: None
</details>

### PR DESCRIPTION
**Pull request by sdementen**
- [x ] closes #15240 
 - [] tests added / passed
 - [x] passes ``git diff upstream/master | flake8 --diff``
 - [x] whatsnew entry

### CODE REVIEW

File: doc/source/whatsnew/v0.20.0.txt [hunk 1]
```diff
@@ -416,6 +416,7 @@ Performance Improvements
 Bug Fixes
 ~~~~~~~~~
 
+- Bug in ``Timestamp.replace`` wrong exception raised when wrong argument name given (:issue:`15240`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.20.0.txt]: say should be TypeError

Comment (jreback) [inline doc/source/whatsnew/v0.20.0.txt]: move to Other API changes section

**Comment by jreback** [comment]
lgtm. ping on green.

File: doc/source/whatsnew/v0.20.0.txt (hunk 1 above)
Comment (sdementen) [inline doc/source/whatsnew/v0.20.0.txt]: done and left in bug fixes as it is a bug introduced by a previous change (not really an excepted API change)

**Comment by sdementen** [comment]
@jreback I am bit puzzle by the errors in the tests. I see issues but I think they are unrelated to my changes. Essentially, I have a) change a signature of function, b) renamed an internal variable (tzinfo) and c) adapted some internal code) but I get error about infinite recursion in some totally unrelated tests (but I may be wrong in my interpretation)

**Comment by jreback** [comment]
@sdementen I pushed a fix to your branch. This is actually a tricky thing, ``.replace`` accepts ``None`` as an actual value for ``tzinfo``. So you need to distinguish between not passing it and passing it with a value of ``None``. The trick is to default it like ``tzinfo=object``, then check if ``tzinfo is object``, means it was not passed.

This shows up in the way it did because ``.replace`` is used a lot, for example, to reset the time zone on objects, e.g. ``Timestamp.replace(tzinfo=None)`` is common in ``dateutil``

**Comment by jreback** [comment]
thanks @sdementen !

**Comment by sdementen** [comment]
Thanks to you for your support in this PR !

**Comment by sdementen** [comment]
BTW, the replace function when used to change the time  (not the tzinfo  part), leads sometimes to inconsistent times when dst is involved in the case the replaced date cross the dst  (and so should have his offset changed). 
Is this a bug ? Or a difficult to solve limit ? Is the introduction of the fold argument a part of a future solution ?
If it is a bug not yet identified, i can add an issue

**Comment by jreback** [comment]
@sdementen you would have to show an example. ``.replace`` replaces current times; however it is possible to then generate an invalid time. ``datetime.datetime`` doesn't handle this at all. ``Timestamp`` *does* to some extant. Look thru open issues and you will see a couple that are out there.

**Pull request closed without merging**